filter.go

```diff
@@ -8,60 +8,126 @@ import (
 	manet "github.com/multiformats/go-multiaddr-net"
 )
 
+type filterEntry struct {
+	f      *net.IPNet
+	reject bool
+}
+
+// Filters is a structure representing a collection of allow/deny
+// net.IPNet filters, together with the RejectByDefault flag, which
+// represents the default filter policy.
+//
+// Note that the last policy added to the Filters is authoritative.
 type Filters struct {
-	mu      sync.RWMutex
-	filters map[string]*net.IPNet
+	mu              sync.RWMutex
+	RejectByDefault bool
+	filters         []*filterEntry
 }
 
+// NewFilters constructs and returns a new set of net.IPNet filters.
+// By default, the new filter rejects no addresses.
 func NewFilters() *Filters {
 	return &Filters{
-		filters: make(map[string]*net.IPNet),
+		RejectByDefault: false,
+		filters:         make([]*filterEntry, 0),
 	}
 }
 
+// AddDialFilter adds a reject rule to the given Filters.  Hosts
+// matching the given net.IPNet filter will be rejected, unless
+// another rule is added which states that they should be accepted.
+//
+// No effort is made to prevent duplication of filters, or to simplify
+// the filters list.
 func (fs *Filters) AddDialFilter(f *net.IPNet) {
 	fs.mu.Lock()
 	defer fs.mu.Unlock()
-	fs.filters[f.String()] = f
+	fs.filters = append(fs.filters, &filterEntry{f: f, reject: true})
+}
+
+// AddAllowFilter adds an accept rule to the given Filters. Hosts
+// matching the given net.IPNet filter will be accepted, unless
+// another policy is added which states that they should be rejected.
+//
+// No effort is made to prevent duplication of filters, or to simplify
+// the filters list.
+func (fs *Filters) AddAllowFilter(f *net.IPNet) {
+	fs.mu.Lock()
+	defer fs.mu.Unlock()
+	fs.filters = append(fs.filters, &filterEntry{f: f, reject: false})
 }
 
+// AddrBlocked parses a ma.Multiaddr, and if it can get a valid netip
+// back applies the Filters, returning true if the given address
+// should be rejected, and false if the given address is allowed.
+//
+// If a parsing error occurs, or no filter matches, the Filters
+// default is returned.
 func (f *Filters) AddrBlocked(a ma.Multiaddr) bool {
 	maddr := ma.Split(a)
 	if len(maddr) == 0 {
-		return false
+		return f.RejectByDefault
 	}
 	netaddr, err := manet.ToNetAddr(maddr[0])
 	if err != nil {
 		// if we cant parse it, its probably not blocked
-		return false
+		return f.RejectByDefault
 	}
 	netip := net.ParseIP(netaddr.String())
 	if netip == nil {
-		return false
+		return f.RejectByDefault
 	}
 
 	f.mu.RLock()
 	defer f.mu.RUnlock()
+
+	var reject bool = f.RejectByDefault
```

Comment:
Right. So we set the default result (`f.RejectByDefault`), then iterate through all filters. The filters slice is in insertion order, so the first filter inserted is applied first. We apply all filters to the address, and if the filter matches the address we change the `reject` outcome to be the outcome specified by the matching rule.

To be specific about the expected behavior:

```
DEFAULT DENY
ALLOW 192.168.0.0/16
DENY    192.160.0.0/24
ALLOW 192.168.0.5/32
```

These rules specify to reject all connections, allowing a `/16`, blocking a `/24` out of that, and and making exception for a single `/32`. These filters work because they're insertion ordered. This how other ACL systems tend to operate with the last rule to match winning. Better ACL systems make some attempt to sort, compact & optimize the ACL table, but this is usually Good Enough.

So as an exercise, given an attempted connection to `192.168.0.5`, we will first take the default outcome which is `DENY`. Then we apply the first rule, which matches and says `ALLOW` presumably because that `/16` is trusted address space. But there may be other relevant policies so we have to keep going. We apply the next rule, which also matches and produces a `DENY`. But another rule could express an exception so we keep going, and we find that the particular `/32` is `ALLOW`ed explicitly.

We can't go back to a map because maps aren't insertion ordered. The unordered map system works fine when you just want to see if there's a reject rule which matches and there's no possibility of countermanding a reject decision once one has been made. However because the map has no concept of ordering it's impossible to impose a hierarchy between rules allowing for exceptions as in this example. It's also not possible to achieve this sort of behavior with two maps (which was the second thing I tried) because again the ordering is ambiguous as to whether the `ALLOW` or `DENY` directive should win given even a simple conflict.